File: core-render-ohos/src/main/cpp/libohos_render/expand/modules/sftp/SftpErrorFormatter.cpp
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "SftpErrorFormatter.h"
#include <sstream>

namespace kuikly {
namespace module {

std::string SftpErrorFormatter::Format(const std::exception &e) {
    std::string msg = e.what();
    // 错误码约定与其它端保持一致（SftpErrorCode）：
    //   9999 NOT_IMPLEMENTED（桩实现必须显式失败，绝不伪报成功）
    //   2001 PERMISSION_DENIED / 2003 NO_SUCH_FILE / 1001-1003 连接类
    int code = 0;
    if (msg.find("not implemented") != std::string::npos) {
        code = 9999;
    } else if (msg.find("NoSuch") != std::string::npos || msg.find("no such") != std::string::npos) {
        code = 2003;
    } else if (msg.find("Permission") != std::string::npos || msg.find("permission") != std::string::npos) {
        code = 2001;
    } else if (msg.find("connect") != std::string::npos) {
        code = 1001;
    } else if (msg.find("auth") != std::string::npos) {
        code = 1003;
    }
    return Format(code, msg);
}

std::string SftpErrorFormatter::Format(int code, const std::string &msg, const std::string &detail) {
    std::ostringstream oss;
    oss << "{\"code\":" << code << ",\"msg\":\"" << msg << "\"";
    if (!detail.empty()) {
        oss << ",\"detail\":\"" << detail << "\"";
    }
    oss << "}";
    return oss.str();
}

}  // namespace module
}  // namespace kuikly

File: core-render-ios/Extension/Modules/SftpErrorFormatter.m
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#import "SftpErrorFormatter.h"

@implementation SftpErrorFormatter

+ (NSString *)formatException:(NSException *)e {
    int code = 0;
    NSString *msg = e.reason ?: @"未知错误";
    NSString *name = e.name ?: @"";
    if ([name containsString:@"ConnectException"] || [name containsString:@"SftpConnectException"]) code = 1001;
    else if ([name containsString:@"Auth"] || [name containsString:@"SftpAuthException"]) code = 1003;
    else if ([name containsString:@"HostKey"]) code = 1004;
    else if ([name containsString:@"Timeout"]) code = 1002;
    else if ([name containsString:@"NoSuch"]) code = 2003;      // NO_SUCH_FILE
    else if ([name containsString:@"Permission"]) code = 2001;  // PERMISSION_DENIED
    else if ([name containsString:@"NotImplemented"]) code = 9999;
    else code = 0;

    NSDictionary *json = @{
        @"code": @(code),
        @"msg": msg,
        @"detail": e.userInfo[@"detail"] ?: [NSNull null]
    };
    NSError *err;
    NSData *data = [NSJSONSerialization dataWithJSONObject:json options:0 error:&err];
    return data ? [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding] : @"{}";
}

+ (NSString *)formatNSError:(NSError *)error {
    int code = 0;
    NSString *msg = error.localizedDescription ?: @"未知错误";
    // NMSSH/NMSFTP errors
    if ([error.domain containsString:@"NMSSH"]) {
        if (error.code == 1001) code = 1001;  // connection
        else if (error.code == 1002) code = 1003;  // auth
        else code = 2999;
    }
    NSDictionary *json = @{
        @"code": @(code),
        @"msg": msg
    };
    NSError *err;
    NSData *data = [NSJSONSerialization dataWithJSONObject:json options:0 error:&err];
    return data ? [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding] : @"{}";
}

@end
